src/riak_kv_qry_buffers_ldb.erl

```diff
@@ -0,0 +1,173 @@
+    ok = eleveldb:close(LdbRef),
+    _ = os:cmd(
+          fmt("rm -rf '~s'", [filename:join(Root, Table)])),
+    ok.
+
+
+-spec add_rows(eleveldb:db_ref(), [riak_kv_qry_buffers:data_row()], integer(),
+               [pos_integer()], [{asc|desc, nulls_first|nulls_last}]) ->
+                      ok | {error, ldb_put_failed}.
+add_rows(LdbRef, Rows, ChunkId,
+         KeyFieldPositions,
+         OrdByFieldQualifiers) ->
+    %% 0. The new key is composed from fields appearing in the ORDER
+    %%    BY clause, and may therefore not work out to be unique. We
+    %%    now index the rows in the chunk to preserve the original
+    %%    order (because imposing any other order is even worse)
+    RowsIndexed = lists:zip(Rows, lists:seq(1, length(Rows))),
+    try
+        lists:foreach(
+          fun({Row, Idx}) ->
+                  %% a. Form a new key from ORDER BY fields
+                  KeyRaw = [lists:nth(Pos, Row) || Pos <- KeyFieldPositions],
+                  %% b. Negate values in columns marked as DESC in ORDER BY clause
+                  KeyOrd = [maybe_negate(F, Dir, NullsGroup)
+                            || {F, {Dir, NullsGroup}} <- lists:zip(KeyRaw, OrdByFieldQualifiers)],
+                  %% c. Combine with chunk id and row idx to ensure uniqueness, and encode.
+                  KeyEnc = sext:encode({KeyOrd, ChunkId, Idx}),
+                  %% d. Encode the record (don't bother constructing a
+                  %%    riak_object with metadata, vclocks):
+                  RowEnc = sext:encode(Row),
+                  ok = eleveldb:put(LdbRef, KeyEnc, RowEnc, [{sync, true}])
+          end,
+          RowsIndexed)
+    catch
+        error:badmatch ->
+            {error, ldb_put_failed}
+    end.
+
+maybe_negate([], desc, nulls_first) ->
+    0;     %% sort before entupled value
+maybe_negate([], asc, nulls_last) ->
+    <<>>;  %% sort after entupled value
+maybe_negate([], desc, nulls_last) ->
+    <<>>;
+maybe_negate([], asc, nulls_first) ->
+    0;
+maybe_negate(F, asc, _) ->
+    {F};
+maybe_negate(F, desc, _) when is_number(F) ->
+    {-F};
+maybe_negate(F, desc, _) when is_binary(F) ->
+    {[<<bnot X>> || <<X>> <= F]};
+maybe_negate(F, desc, _) when is_boolean(F) ->
+    {not F}.
+
+
+
+-spec fetch_rows(eleveldb:db_ref(), undefined|non_neg_integer(), undefined|pos_integer()) ->
+                        {ok, [riak_kv_qry_buffers:data_row()]} | {error, term()}.
+fetch_rows(LdbRef, undefined, Limit) ->
+    fetch_rows(LdbRef, 0, Limit);
+fetch_rows(LdbRef, Offset, LimitOrUndef) ->
+    %% Because the database is read-only, the plan is to keep a cache
+    %% of iterators for faster seeking (to the nearest stored),
+    %% ideally also enable eleveldb to do the folding from stored
+    %% iterators rather than from 'first'.
+    FetchLimitFn =
+        fun(_KV, {Off, Lim, Pos, Acc}) when Pos < Off ->
+                {Off, Lim, Pos + 1, Acc};
+           %% Fetching K (let alone V) while "seeking" to our Offset
```

Comment:
@hmmr  API for manual manipulation of iterators within Erlang exists.  See riak_core/hashtree.erl for examples.  That code is already doing much of what you are requesting.